### PR TITLE
Add function to handle opensuse-welcome 'don't show on next startup'

### PR DESCRIPTION
**Pull request by Zaoliang**
see https://progress.opensuse.org/issues/56777
verification test:
http://f40.suse.de/tests/5286#step/installation/13

### CODE REVIEW

**Comment by SergioAtSUSE** [comment]
@okurz, the module was a bad idea. But, I didn't realize until we got the failures on other scenarios:

like
- http://f40.suse.de/tests/5270#step/installation/10
- https://openqa.opensuse.org/tests/1076776#step/opensuse_welcome/2

Which means that it would need to be scheduled and return on those scenarios (dirty)
try to not schedule on them (dirty)
or not schedule it at all. The first module that logs in should handle the popup: untick the checkbox and close it forever (for this user).
In autoyast scenario that would be `installation` module. In root-only scenario nothing is needed. In normal scenario that would be `first_boot module`.
And because that module cannot be schedule for autoyast scenario, it has to be a function.

Also, the current approach causes a lot of microcode on each graphic module:

- http://f40.suse.de/tests/5270#step/autoyast_reboot/1
- http://f40.suse.de/tests/5270#step/first_boot/1
- http://f40.suse.de/tests/5270#step/opensuse_welcome/1

Not only dirty, not only duplicate. Leaving that popup in the background is dangerous. Each new module that is added to those scenarios would need to have this assert_screen and pass and having that windows will hit us (remember gnucash and the problems we had to close the popups. The X button is the same so assert_screen is a Russian roulette, and alt-f4 doesn't work when the openQA ghost focus the wrong window.)

Please, see the whole epic. To avoid duplication, we first need the function, then remove the module opensuse_welcome, then remove all the "worked around" module looking for opensuse-welcome needle.
https://progress.opensuse.org/issues/56267

File: tests/autoyast/installation.pm [hunk 1]
```diff
@@ -343,7 +343,8 @@ sub run {
             return;
         }
         elsif (match_has_tag('opensuse-welcome')) {
-            return;
+            assert_and_click 'dont-show-on-next-startup';
+            assert_and_click 'close-welcome';
```
Comment (SergioAtSUSE) [inline tests/autoyast/installation.pm]: This should be in a function to be reused.

For example, here: https://openqa.opensuse.org/tests/1077390#step/multi_users_dm/16

**Comment by okurz** [comment]
Please involve the people that previously worked on the welcome module in this, no problem to rediscuss the existing approach. But I did not see that this PR would correct the old approach but rather add another unrelated way without even mentioning the other, so I was not aware that you already knew all the implications of the existing approach

**Comment by SergioAtSUSE** [comment]
CC @StefanBruens 
:wink:

File: tests/autoyast/installation.pm [hunk 2]
```diff
@@ -122,6 +122,11 @@ sub verify_timeout_and_check_screen {
     return check_screen $needles, $check_time;
 }
 
+sub untick_welcome_on_next_startup {
+    assert_and_click 'dont-show-on-next-startup';
+    assert_and_click 'close-welcome';
+}
```
Comment (SergioAtSUSE) [inline tests/autoyast/installation.pm]: To be reusable, this should be in a shared lib.
I think the best match would be **lib/x11utils.pm**

**Comment by Soulofdestiny** [comment]
> Please involve the people that previously worked on the welcome module in this, no problem to rediscuss the existing approach. But I did not see that this PR would correct the old approach but rather add another unrelated way without even mentioning the other, so I was not aware that you already knew all the implications of the existing approach

can you explain your comment here? what do you not like about this? that it is split in different tasks and we will do it step by step?

**Comment by okurz** [comment]
> can you explain your comment here? what do you not like about this? that it is split in different tasks and we will do it step by step?

I am sorry if I missed something. The [description](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/8864#issue-338042385) did not really reveal anything and the subject only states what is done, not anything pointing to a reason. I am not complaining, just stating what is not obvious to me. And as one of the main challenges within os-autoinst-distri-opensuse I see "stepping on each other's toes" and "unexpected impacts in other products/versions" so I just want to ensure we at least considered that carefully.

> iiuc, all relevant people were involved and all possible scenarios are covered.

I also did not see that so I just kindly asked. So you have also included ggardet and dimstar? Just to make sure:

@DimStar77 @ggardet please take a look as you have created/edited/reviewed the code for the openSUSE welcome handling as well, e.g. in tests/installation/opensuse_welcome.pm

> haven't you always been in favor of epics?

I don't understand what you mean with this. A pull request can not be an "epic" and the mentioned progress ticket is also not an epic. Maybe you just mean the comment by Sergio.

**Comment by ggardet** [comment]
It will likely fail. Please check the `opensuse_welcome.pm` code. We need to handle some retries.
I would suggest to move the code from `opensuse_welcome.pm` to a function and call that function from `opensuse_welcome.pm` and any other place where it is relevant.

Please schedule verification run on o3 and for aarch64, please.

**Comment by SergioAtSUSE** [comment]
@ggardet, please see https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/8864#issuecomment-551152772
And the epic: https://progress.opensuse.org/issues/56267

The opensuse-welcome module doesn't scale and increase the maintenance time.
This PR only takes care of autoyast installation.
The other scenarios that are going to be handled in other tickets are:
- root-only
- multi users
- wait_install+first_boot combination

**Comment by SergioAtSUSE** [comment]
@Zaoliang, please, try 100 runs on aarch64

**Comment by ggardet** [comment]
@SergioAtSUSE sorry, but you _need_ to move/reuse the code from `opensuse_welcome.pm` since it has been tested and is reliable. Do not start from scratch as it will not works with just 2 `assert_and_click`.

**Comment by StefanBruens** [comment]
> Also, the current approach causes a lot of microcode on each graphic module:
>     * http://f40.suse.de/tests/5270#step/autoyast_reboot/1
>     * http://f40.suse.de/tests/5270#step/first_boot/1
>     * http://f40.suse.de/tests/5270#step/opensuse_welcome/1

Please use links everyone can use ...

**Comment by coolo** [comment]
That's easier said than done.

**Comment by ggardet** [comment]
> That's easier said than done.

You can schedule easily on o3, but please do not schedule 100 jobs for aarch64. ;)

**Comment by StefanBruens** [comment]
IMHO a good first step to remove any duplication would be to make autoyast/installation.pm more similar to  installation/await_install.pm.

autoyast/installation.pm is a code monster, it not only handles installation, but everything including the first login. This implicitly includes dealing with opensuse_welcome.

When autoyast/installation.pm would exit like await_install, i.e. when the "Reboot now" dialog appears, any opensuse_welcome handling in installation.pm would be moot.

**Comment by SergioAtSUSE** [comment]
> You can schedule easily on o3, but please do not schedule 100 jobs for aarch64. ;)

AFAIK openqa-clone-custom-git-refspec cannot take code changes and needles: https://github.com/os-autoinst/os-autoinst-needles-opensuse/pull/607/files?short_path=e540152#diff-e540152edd8180eb728fde4330a4a04a

**Comment by ggardet** [comment]
> > You can schedule easily on o3, but please do not schedule 100 jobs for aarch64. ;)
> 
> AFAIK openqa-clone-custom-git-refspec cannot take code changes and needles: 

You should not need new needles, the existing ones should be enough.

**Comment by SergioAtSUSE** [comment]
> IMHO a good first step to remove any duplication would be to make autoyast/installation.pm more similar to installation/await_install.pm.

This is not the only case where the **opensuse_welcome** module is causing more effort than the problem is tries to solve:

- root-only: https://openqa.opensuse.org/tests/1076776#step/opensuse_welcome/2
- change-password: https://openqa.opensuse.org/tests/1030211#step/change_password/51
- multi-users: https://openqa.opensuse.org/tests/1077390#step/multi_users_dm/16

> autoyast/installation.pm is a code monster, it not only handles installation, but everything including the first login.
> This implicitly includes dealing with opensuse_welcome.
> When autoyast/installation.pm would exit like await_install, i.e. when the "Reboot now" dialog appears, any opensuse_welcome handling in installation.pm would be moot.

This scenario and module are maintained by other team. Any team is free to handle their scenarios and module how they wish, until they create something that affects other teams.
Please discuss this topic with them.
The module installation is not affecting us, but the solution for the opensuse welcome is affecting us all.

**Comment by ggardet** [comment]
It looks better like this. Thanks.

**Comment by SergioAtSUSE** [comment]
@Zaoliang, I realized that there is no testsuite autoyast_gnome for aarch64 on O3.
I have just enabled it in the development job group: https://openqa.opensuse.org/admin/job_templates/38
On Monday we will be able to clone it. (I assume we will get a new build by then)

**Comment by okurz** [comment]
Sergio your "team responsibility" story is ... sorry to be blunt: Bullshit. Walk 5m to talk to "them" or just ask them if they agree with changes in a PR

**Comment by Zaoliang** [comment]
autoyast_gnome on aarch64:

http://f40.suse.de/tests/5290#step/installation/14 shows successful test run.

**Comment by ggardet** [comment]
`autoyast_gnome` on `aarch64` without this PR is available on o3 in `Development > Tumbleweed`: https://openqa.opensuse.org/tests/1083353

It will be in regular Tumbleweed aarch64 with next snapshot.

**Comment by Zaoliang** [comment]
openqa-clone-custom-git-refspec doesn't work at moment. So I cannot provide verification test run with this PR on o3.

**Comment by ggardet** [comment]
> openqa-clone-custom-git-refspec doesn't work at moment. So I cannot provide verification test run with this PR on o3.

Here it is: https://openqa.opensuse.org/tests/1083422
(with a hotpatched `openqa-clone-custom-git-refspec`: https://github.com/os-autoinst/openQA/pull/2475/files )

**Comment by ggardet** [comment]
VR on aarch64 Tumbleweed on o3 failed: https://openqa.opensuse.org/tests/1083422#step/installation/15

You may need to wait for `opensuse_welcome` explicitly when applicable, to ensure your system is fully booted.

**Comment by Zaoliang** [comment]
successful verification test for autoyast_gnome on osd:
https://openqa.opensuse.org/tests/1083432#step/installation/17

**Comment by ggardet** [comment]
> successful verification test for autoyast_gnome on osd:
> https://openqa.opensuse.org/tests/1083432#step/installation/17

It failed because you need to drop the `opensuse_welcome` module now.

**Comment by Zaoliang** [comment]
okay, here is the test on osd without opensuse_welcome (this will be handled in another ticket).

https://openqa.opensuse.org/tests/1084333

**Comment by ggardet** [comment]
It looks good. Let's see how it behaves on o3 on long term. :)

**Comment by SergioAtSUSE** [comment]
Test suite updated.

**Pull request merged**